### PR TITLE
QskArcShadownode

### PR DESCRIPTION
**Pull request by rick-vogel**
|||
|-|-|
|Recording 2023-12-09 at 16 19 40|image|

### CODE REVIEW

File: src/nodes/QskArcNode.cpp [hunk 1]
```diff
@@ -89,29 +99,44 @@ void QskArcNode::setArcData( const QRectF& rect, const QskArcMetrics& arcMetrics
...
+            shadowNode = new QskArcShadowNode;
+            QskSGNode::setNodeRole( shadowNode, ShadowRole );            
+        }
+        
+        shadowNode->update( arcRect, metrics, shadowColor, shadowMetrics, borderWidth );
+    }
+    else
+    {
+        delete shadowNode;
+        shadowNode = nullptr;
```
Comment (rick-vogel) [inline src/nodes/QskArcNode.cpp:123]: @uwerat Without `delete` dangling pointer ⚠️

File: src/nodes/QskArcNode.cpp [hunk 2]
```diff
@@ -132,5 +155,23 @@ void QskArcNode::setArcData( const QRectF& rect, const QskArcMetrics& arcMetrics
...
+    if ( oldChildCount != newChildCount ) 
+    {
+        removeAllChildNodes();
+        
+        for ( QSGNode* node : { ( QSGNode* ) shadowNode, ( QSGNode* ) fillNode, ( QSGNode* ) borderNode })
+        {
+            if ( node != nullptr ) 
+            {
+                appendChildNode( node );
+            }
```
Comment (rick-vogel) [inline src/nodes/QskArcNode.cpp:174]: @uwerat Is it too costly to reattach all children in the right order? Makes the implementation easy and extendable but if you want I can revert it to prepend, append and insertChildNodeAfter/Before ...

File: src/nodes/QskArcShadowNode.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+class QskArcShadowNodePrivate;
+
+class QskArcShadowNode : public QSGGeometryNode
+{
+  public:
+    QskArcShadowNode();
+    ~QskArcShadowNode() override;
+
+    void update( const QRectF& rect, const QskArcMetrics& arcMetrics, const QColor& shadowColor,
+        const QskShadowMetrics& shadowMetrics, qreal borderWidth );
```
Comment (rick-vogel) [inline src/nodes/QskArcShadowNode.h:24]: @uwerat Currently only QskShadowMetrics::offset and QskShadowMetrics::spreadRadius are being used. Maybe the shadow characteristics need to be synced.

Comment (rick-vogel) [reply src/nodes/QskArcShadowNode.h:24]: Now all `QskShadow` properties are being used like for `QskBoxNodes`

File: src/nodes/QskArcNode.cpp (hunk 2 above)
Comment (rick-vogel) [reply src/nodes/QskArcNode.cpp:174]: Adapted to `QskBoxNode`'s pattern

File: src/nodes/QskArcNode.cpp [hunk 4]
```diff
@@ -89,29 +99,44 @@ void QskArcNode::setArcData( const QRectF& rect, const QskArcMetrics& arcMetrics
...
+        delete shadowNode;
         delete fillNode;
         delete borderNode;
-
         return;
     }
 
     const auto path = metrics.painterPath( arcRect );
 
+    if ( shadowColor.alpha() > 0.0 )
```
[the remark this answers is not in the document]
Comment (rick-vogel) [reply src/nodes/QskArcNode.cpp]: Now the shadow's visibility depends on the fill node's visibility

File: src/nodes/QskArcShadowNode.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    }
+}
+
+void QskArcShadowNode::update( const QRectF& rect, const QskArcMetrics& arcMetrics,
+    const QColor& shadowColor, const QskShadowMetrics& shadowMetrics, const qreal borderWidth )
+{
+    Q_D( QskArcShadowNode );
+    
+    const auto w = qMax(0.0, borderWidth);
+    const auto shadowRect = rect.adjusted(-w, -w, +w, +w);
```
Comment (rick-vogel) [inline src/nodes/QskArcShadowNode.cpp:316]: add spread radius + blur radius to `w`

**Comment by uwerat** [comment]
Code went into a feature branch: https://github.com/uwerat/qskinny/tree/features/arcshadow to do the final adjustment

**Pull request closed without merging**